src/cunumeric/sort/sort.cu

```diff
@@ -95,338 +137,1460 @@ struct SortPiece {
+      recv_right_data.values.destroy();
+    }
+  }
+}
+
+/////////////////////////////////////////////////////////////////////////////////////////////////
+/////////////////////////////////////////////////////////////////////////////////////////////////
+/////////////////////////////////////////////////////////////////////////////////////////////////
+/////////////////////////////////////////////////////////////////////////////////////////////////
+/////////////////////////////////////////////////////////////////////////////////////////////////
+/////////////////////////////////////////////////////////////////////////////////////////////////
+
+template <typename VAL>
+void sample_sort_nccl_nd(SortPiece<VAL> local_sorted,
+                         void* output_ptr,
+                         /* global domain information */
+                         size_t my_rank,  // global NCCL rank
+                         size_t num_ranks,
+                         size_t segment_size_g,
+                         /* domain information in sort dimension */
+                         size_t my_sort_rank,    // local rank id in sort dimension
+                         size_t num_sort_ranks,  // #ranks that share a sort dimension
+                         size_t* sort_ranks,     // rank ids that share a sort dimension with us
+                         size_t segment_size_l,  // (local) segment size
+                         /* other */
+                         bool argsort,
+                         cudaStream_t stream,
+                         ncclComm_t* comm)
 {
   size_t volume = local_sorted.size;
 
-  // collect local samples - for now we take num_ranks samples for every node
-  // worst case this leads to 2*N/ranks elements on a single node
-  size_t num_local_samples = num_ranks;
+  /////////////////////////////////////////////////////////////////////////////////////////////////
+  /////////////// Part 0: detection of empty nodes
+  /////////////////////////////////////////////////////////////////////////////////////////////////
+
+  // first of all we need to check for processes that don't want
+  // to take part in the computation. This might lead to a reduction of
+  // sort ranks. Note that if segment_size_l>0 && volume==0 means that we have
+  // a full sort group being empty, this should not affect local sort rank size.
+  {
+    auto worker_count_d = create_buffer<int32_t>(1, Memory::GPU_FB_MEM);
+    int worker_count[1] = {(segment_size_l > 0 ? 1 : 0)};
+    CHECK_CUDA(cudaMemcpyAsync(
+      worker_count_d.ptr(0), &worker_count[0], sizeof(int32_t), cudaMemcpyHostToDevice, stream));
+    CHECK_NCCL(ncclAllReduce(
+      worker_count_d.ptr(0), worker_count_d.ptr(0), 1, ncclInt32, ncclSum, *comm, stream));
+    CHECK_CUDA(cudaMemcpyAsync(
+      &worker_count[0], worker_count_d.ptr(0), sizeof(int32_t), cudaMemcpyDeviceToHost, stream));
+    CHECK_CUDA(cudaStreamSynchronize(stream));
+    if (worker_count[0] < num_ranks) {
```

Comment:
Yes, there are still some unwritten dependencies here. In case we want to allow more dynamic resource assignment we would need to provide some more information on it to the tasks I suppose.